plasmapy/atomic/ionization_states.py

```diff
@@ -0,0 +1,965 @@
+
+from .atomic import atomic_number
+from .particle_class import Particle
+from .particle_input import particle_input
+from .symbols import particle_symbol
+from ..utils import (AtomicError, ChargeError, InvalidParticleError, check_quantity)
+
+State = collections.namedtuple(
+    'State', [
+        'integer_charge',
+        'ionic_fraction',
+        'ionic_symbol',
+    ])
+
+_number_density_errmsg = (
+    "Number densities must be Quantity objects with units of inverse "
+    "volume."
+)
+
+
+class IonizationState:
+    """
+    Representation of the ionization state distribution of a single
+    element or isotope.
+
+    Parameters
+    ----------
+    particle: str, int, np.integer, or ~plasmapy.atomic.Particle
+        A `str` or `~plasmapy.atomic.Particle` instance representing
+        an element or isotope, or an `int` representing the atomic
+        number of an element.
+
+    ionic_fractions: ~numpy.ndarray, list, tuple, or ~astropy.units.Quantity
+        The ionization fractions of an element, where the indices
+        correspond to integer charge.  This argument should contain the
+        atomic number plus one items, and must sum to one within an
+        absolute tolerance of `tol` if dimensionless.  Alternatively,
+        this argument may be a `~astropy.units.Quantity` that represents
+        the number densities of each neutral/ion.
+
+    T_e: ~astropy.units.Quantity, keyword-only, optional
+        The electron temperature or thermal energy per particle.
+
+    n_e: ~astropy.units.Quantity, keyword-only, optional
+        The electron number density.
+
+    n_elem: ~astropy.units.Quantity, keyword-only, optional
+        The number density of the element, including neutrals and all
+        ions.
+
+    tol: float or int, keyword-only, optional
+        The absolute tolerance used by `~numpy.isclose` when testing
+        normalizations and making comparisons.  Defaults to `1e-15`.
+
+    Raises
+    ------
+    ~plasmapy.utils.AtomicError
+        If the ionic fractions are not normalized or contain invalid
+        values
+
+    ~plasmapy.utils.InvalidParticleError
+        If the particle is invalid.
+
+    Examples
+    --------
+    >>> states = IonizationState('H', [0.6, 0.4], n_elem=1*u.cm**-3, T_e=11000*u.K)
+    >>> states.ionic_fractions
+    array([0.6, 0.4])
+    >>> states.n_e  # electron number density
+    <Quantity 400000. 1 / m3>
+    >>> states.n_elem  # element number density
+    <Quantity 1000000. 1 / m3>
+
+    Notes
+    -----
+    Only one of `n_e` and `n_elem` may be set.
+
+    """
+
+    # TODO: Allow this class to (optionally?) handle negatively charged
+    # ions.  There are instances where singly negatively charged ions
+    # are important in astrophysical plasmas, such as H- in the
+    # atmospheres of relatively cool stars.  There may be some rare
+    # situations where doubly negatively charged ions show up too,
+    # though triply negatively charged ions are very unlikely.
```

Comment:
What would it take to allow this? A list like `positive` = [True, False]`? Or would you need to provide, instead of `atomic_number +1`, `2atomic_number + 1` floats to the second argument?